pkg/cloudprovider/vsphereparavirtual/vmoperator/client/virtualmachine_client_test.go

```diff
@@ -0,0 +1,325 @@
+			},
+			expectedErr: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			_, err := vms.Create(context.Background(), testCase.virtualMachine, metav1.CreateOptions{})
+			assert.NoError(t, err)
+			if testCase.deleteFunc != nil {
+				fc.PrependReactor("delete", "*", testCase.deleteFunc)
+			}
+			err = vms.Delete(context.Background(), testCase.virtualMachine.Name, metav1.DeleteOptions{})
+			if testCase.expectedErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestVMGet(t *testing.T) {
+	vms, fc := initVMTest()
+	testCases := []struct {
+		name           string
+		virtualMachine *vmopv1alpha1.VirtualMachine
+		getFunc        func(action clientgotesting.Action) (handled bool, ret runtime.Object, err error)
+		expectedVM     *vmopv1alpha1.VirtualMachine
+		expectedErr    bool
+	}{
+		{
+			name: "Get: when everything is ok",
+			virtualMachine: &vmopv1alpha1.VirtualMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-vm",
+				},
+				Spec: vmopv1alpha1.VirtualMachineSpec{
+					ImageName: "test-image",
+				},
+			},
+			expectedVM: &vmopv1alpha1.VirtualMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-vm",
+				},
+				Spec: vmopv1alpha1.VirtualMachineSpec{
+					ImageName: "test-image",
+				},
+			},
+		},
+		{
+			name: "Get: when get error",
+			virtualMachine: &vmopv1alpha1.VirtualMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-vm-error",
+				},
+				Spec: vmopv1alpha1.VirtualMachineSpec{
+					ImageName: "test-image",
+				},
+			},
+			getFunc: func(action clientgotesting.Action) (bool, runtime.Object, error) {
+				return true, nil, fmt.Errorf("test error")
+			},
+			expectedVM:  nil,
+			expectedErr: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			_, err := vms.Create(context.Background(), testCase.virtualMachine, metav1.CreateOptions{})
+			assert.NoError(t, err)
+			if testCase.getFunc != nil {
+				fc.PrependReactor("get", "*", testCase.getFunc)
+			}
+			actualVM, err := vms.Get(context.Background(), testCase.virtualMachine.Name, metav1.GetOptions{})
+			if testCase.expectedErr {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+				assert.Equal(t, testCase.expectedVM.Spec, actualVM.Spec)
+			}
+		})
+	}
+}
+
+func TestVMList(t *testing.T) {
+	vms, fc := initVMTest()
+	testCases := []struct {
+		name           string
+		virtualMachine *vmopv1alpha1.VirtualMachine
+		listFunc       func(action clientgotesting.Action) (handled bool, ret runtime.Object, err error)
+		expectedVMNum  int
+		expectedErr    bool
+	}{
+		{
+			name: "List: when everything is ok",
+			virtualMachine: &vmopv1alpha1.VirtualMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-vm",
+				},
+				Spec: vmopv1alpha1.VirtualMachineSpec{
+					ImageName: "test-image",
+				},
+			},
+			expectedVMNum: 1,
```

Comment:
Maybe let's add a second VM 

and an additional test case for 0 VMs

Reply:
make sense, thanks @sbueringer !